File: go/HPKE.go
/*
 * Copyright (c) 2012-2020 MIRACL UK Ltd.
 *
 * This file is part of MIRACL Core
 * (see https://github.com/miracl/core).
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* Hybrid Public Key Encryption */

/* Following https://datatracker.ietf.org/doc/draft-irtf-cfrg-hpke/?include_text=1 */

package XXX

//import "fmt"
import "github.com/miracl/core/go/core"

func reverse(X []byte) {
	lx:=len(X)
	for i:=0;i<lx/2;i++ {
		ch :=X[i]
		X[i]=X[lx-i-1]
		X[lx-i-1]=ch
	}
}


func labeledExtract(SALT []byte,label string,IKM []byte) []byte {
	rfc:="RFCXX"+"XX "
	prefix:=[]byte(rfc+label)
	var LIKM []byte
	for i:=0;i<len(prefix);i++ {
		LIKM=append(LIKM,prefix[i])
	}
	if (IKM!=nil) {
		for i:=0;i<len(IKM);i++ {
			LIKM=append(LIKM,IKM[i])
		}
	}
	return core.HKDF_Extract(core.MC_SHA2,HASH_TYPE,SALT,LIKM);
}

func labeledExpand(PRK []byte,label string,INFO []byte,L int) []byte {
	rfc:="RFCXX" +"XX "
	prefix:=[]byte(rfc+label)
	ar := core.InttoBytes(L,2)
	var LINFO []byte
	for i:=0;i<len(ar);i++ {
		LINFO = append(LINFO, ar[i])
	}
	for i:=0;i<len(prefix);i++ {
		LINFO=append(LINFO,prefix[i])
	}	
	for i:=0;i<len(INFO);i++ {
		LINFO=append(LINFO,INFO[i])
	}	

	return core.HKDF_Expand(core.MC_SHA2,HASH_TYPE,L,PRK,LINFO)
}

func extractAndExpand(DH []byte,context []byte) []byte {
	PRK:=labeledExtract(nil,"dh",DH);
	return labeledExpand(PRK,"prk",context,HASH_TYPE);
}

func Encap(config_id int,SKE []byte,pkE []byte,pkR []byte) []byte {
	var skE [EGS]byte
	var DH [EFS]byte
	var kemcontext [] byte
	kem := config_id&255

	for i:=0;i<EGS;i++ {
		skE[i]=SKE[i]
	}
	if kem==32 || kem==33 {
		reverse(skE[:])
		if kem==32 {
			skE[EGS-1]&=248
			skE[0]&=127
			skE[0]|=64
		} else {
			skE[EGS-1]&=252
			skE[0]|=128
		}
	}
	
	ECDH_KEY_PAIR_GENERATE(nil, skE[:], pkE)
	if kem==32 || kem==33 {
		reverse(pkR)
	}
	ECDH_ECPSVDP_DH(skE[:], pkR[:], DH[:])
	if kem==32 || kem==33 {
		reverse(pkR)
		reverse(pkE)
		reverse(DH[:])
	}
	for i:=0;i<len(pkE);i++ {
		kemcontext=append(kemcontext,pkE[i]);
	}
	for i:=0;i<len(pkR);i++ {
		kemcontext=append(kemcontext,pkR[i]);
	}

	return extractAndExpand(DH[:],kemcontext)
}
		
func Decap(config_id int,pkE []byte,SKR []byte) []byte {
	var skR [EGS]byte
	var DH [EFS]byte
	pkR:=make([]byte,len(pkE)) 
	var kemcontext [] byte
	kem := config_id&255
	for i:=0;i<EGS;i++ {
		skR[i]=SKR[i]
	}
	if kem==32 || kem==33 {
		reverse(skR[:])
		reverse(pkE)
		if kem==32 {
			skR[EGS-1]&=248
			skR[0]&=127
			skR[0]|=64
		} else {
			skR[EGS-1]&=252
			skR[0]|=128
		}
	}
	ECDH_ECPSVDP_DH(skR[:], pkE, DH[:])
	if kem==32 || kem==33 {
		reverse(pkE)
		reverse(DH[:])
	}
	ECDH_KEY_PAIR_GENERATE(nil,skR[:],pkR)
	if kem==32 || kem==33 {
		reverse(pkR[:])
	}
	for i:=0;i<len(pkE);i++ {
		kemcontext=append(kemcontext,pkE[i]);
	}
	for i:=0;i<len(pkR);i++ {
		kemcontext=append(kemcontext,pkR[i]);
	}

	return extractAndExpand(DH[:],kemcontext)
}

func AuthEncap(config_id int,SKE []byte,pkE []byte,pkR []byte,SKS []byte) []byte {
	var skE [EGS]byte	
	var skS [EGS]byte
	var DH [EFS]byte
	var DH1 [EFS]byte
	pklen:=len(pkE)
	pkS:=make([]byte,pklen) 
	kemcontext:=make([] byte,3*pklen)
	kem := config_id&255

	for i:=0;i<EGS;i++ {
		skE[i]=SKE[i]
		skS[i]=SKS[i]
	}	
	if kem==32 || kem==33 {
		reverse(skE[:])
		reverse(skS[:])
		if kem==32 {
			skE[EGS-1]&=248
			skE[0]&=127
			skE[0]|=64
			skS[EGS-1]&=248
			skS[0]&=127
			skS[0]|=64	
		} else {
			skE[EGS-1]&=252
			skE[0]|=128
			skS[EGS-1]&=252
			skS[0]|=128
		}
	}
	
	ECDH_KEY_PAIR_GENERATE(nil, skE[:], pkE)
	if kem==32 || kem==33 {
		reverse(pkR)
	}
	ECDH_ECPSVDP_DH(skE[:], pkR, DH[:])
	ECDH_ECPSVDP_DH(skS[:], pkR, DH1[:])
	if kem==32 || kem==33 {
		reverse(DH[:])
		reverse(DH1[:])
	}
	var ZZ [2*EFS]byte
	for i:=0;i<EFS;i++ {
		ZZ[i]=DH[i];
		ZZ[EFS+i]=DH1[i];
	}

	ECDH_KEY_PAIR_GENERATE(nil, skS[:], pkS[:])
	
	if kem==32 || kem==33 {
		reverse(pkR)
		reverse(pkE)
		reverse(pkS[:])
	}
	for i:=0;i<pklen;i++ {
		kemcontext[i] = pkE[i]
        kemcontext[pklen+i]= pkR[i]
        kemcontext[2*pklen+i]= pkS[i]
	}
	return extractAndExpand(ZZ[:],kemcontext)
}

func AuthDecap(config_id int,pkE []byte,SKR []byte,pkS []byte) []byte  {
	var skR [EGS]byte	
	var DH [EFS]byte
	var DH1 [EFS]byte
	pklen:=len(pkE)
	pkR:=make([]byte,pklen) 
	kemcontext:=make([] byte,3*pklen)

	kem := config_id&255

	for i:=0;i<EGS;i++ {
		skR[i]=SKR[i]
	}
	if kem==32 || kem==33 {
		reverse(skR[:])
		reverse(pkE)
		reverse(pkS)
		if kem==32 {
			skR[EGS-1]&=248
			skR[0]&=127
			skR[0]|=64
		} else {
			skR[EGS-1]&=252
			skR[0]|=128
		}
	}
	ECDH_ECPSVDP_DH(skR[:], pkE, DH[:])
	ECDH_ECPSVDP_DH(skR[:], pkS, DH1[:])
	if kem==32 || kem==33 {
		reverse(DH[:])
		reverse(DH1[:])
	}
	var ZZ [2*EFS]byte
	for i:=0;i<EFS;i++ {
		ZZ[i]=DH[i];
		ZZ[EFS+i]=DH1[i];
	}

	ECDH_KEY_PAIR_GENERATE(nil, skR[:], pkR[:])

	if kem==32 || kem==33 {
		reverse(pkR[:])
		reverse(pkE)
		reverse(pkS)
	}
	for i:=0;i<pklen;i++ {
		kemcontext[i] = pkE[i]
        kemcontext[pklen+i]= pkR[i]
        kemcontext[2*pklen+i]= pkS[i]
	}
	return extractAndExpand(ZZ[:],kemcontext)
}

/*
func printBinary(array []byte) {
	for i := 0; i < len(array); i++ {
		fmt.Printf("%02x", array[i])
	}
	fmt.Printf("\n")
}
*/

func KeySchedule(config_id int,mode int,Z []byte,info []byte,psk []byte,pskID []byte) ([]byte,[]byte,[]byte) {
	var context []byte

	kem_id:=config_id&255
	kdf_id:=(config_id>>8)&3
	aead_id:=(config_id>>10)&3

	ar:=core.InttoBytes(kem_id, 2)
	for i:=0;i<len(ar);i++ {
		context = append(context, ar[i])
	}
	ar=core.InttoBytes(kdf_id, 2)
	for i:=0;i<len(ar);i++ {
		context = append(context, ar[i])
	}
	ar=core.InttoBytes(aead_id, 2)
	for i:=0;i<len(ar);i++ {
		context = append(context, ar[i])
	}
	ar = core.InttoBytes(mode, 1)
	for i:=0;i<len(ar);i++ {
		context = append(context, ar[i])
	}

	var ZEROS [HASH_TYPE]byte
	for i:=0;i<HASH_TYPE;i++ {
		ZEROS[i]=0
	}

	H:=labeledExtract(ZEROS[:],"pskID_hash",pskID);
	for i:=0;i<HASH_TYPE;i++ {
		context=append(context,H[i])
	}
	H=labeledExtract(ZEROS[:],"info",info)
	for i:=0;i<HASH_TYPE;i++ {
		context=append(context,H[i])
	}

	if psk==nil {
		H=labeledExtract(ZEROS[:],"psk_hash",ZEROS[:])
	} else {
		H=labeledExtract(ZEROS[:],"psk_hash",psk)
	}
	secret:=labeledExtract(H,"zz",Z)

	key:=labeledExpand(secret,"key",context,AESKEY)
	nonce:=labeledExpand(secret,"nonce",context,12)
	exp_secret:=labeledExpand(secret,"exp",context,HASH_TYPE)

	return key,nonce,exp_secret
}	

File: javascript/hpke.js
/*
 * Copyright (c) 2012-2020 MIRACL UK Ltd.
 *
 * This file is part of MIRACL Core
 * (see https://github.com/miracl/core).
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* Hybrid Public Key Encryption */

/* Following https://datatracker.ietf.org/doc/draft-irtf-cfrg-hpke/?include_text=1 */

var HPKE = function(ctx) {
    "use strict";

    var HPKE = {

        reverse: function(X) {
	        var i,ch,lx=X.length;
	        for (i=0;i<lx/2;i++)  {
		        ch =X[i];
		        X[i]=X[lx-i-1];
		        X[lx-i-1]=ch;
	        }
        },

        LabeledExtract: function(SALT,label,IKM) {
            var rfc="RFCXX"+"XX ";
            var prefix=ctx.ECDH.asciitobytes(rfc+label);
            var IKMlen;
            if (IKM==null)
                IKMlen=0;
            else
                IKMlen=IKM.length;
            var LIKM = [];
            var k=0;
            for (var i=0;i<prefix.length;i++)
                LIKM[k++]=prefix[i];
            if (IKM!=null) {
                for (var i=0;i<IKM.length;i++)
                    LIKM[k++]=IKM[i];
            }

            return ctx.HMAC.HKDF_Extract(ctx.HMAC.MC_SHA2,ctx.ECP.HASH_TYPE,SALT,LIKM);
        },

        LabeledExpand: function(PRK,label,INFO,L) {
            var ar = ctx.HMAC.inttobytes(L, 2);
            var rfc="RFCXX" +"XX ";
            var prefix=ctx.ECDH.asciitobytes(rfc+label);   
            var LINFO = [];
            LINFO[0]=ar[0];
            LINFO[1]=ar[1];
            var k=2;
            for (var i=0;i<prefix.length;i++)
                LINFO[k++]=prefix[i];
            for (var i=0;i<INFO.length;i++)
                LINFO[k++]=INFO[i];

            return ctx.HMAC.HKDF_Expand(ctx.HMAC.MC_SHA2,ctx.ECP.HASH_TYPE,L,PRK,LINFO);
        },

        ExtractAndExpand: function(DH,CONTEXT) {
            var PRK=this.LabeledExtract(null,"dh",DH);
            return this.LabeledExpand(PRK,"prk",CONTEXT,ctx.ECP.HASH_TYPE);
        },

        encap: function(config_id,SKE,pkE,pkR) {
            var skE = [];
            var DH = [];
            var KEMCONTEXT=[];
	        var kem = config_id&255;
	        
		    for (var i=0;i<ctx.ECDH.EGS;i++) 
			    skE[i]=SKE[i];
		    if (kem==32 || kem==33) {
			    this.reverse(skE);
			    if (kem==32) {
				    skE[ctx.ECDH.EGS-1]&=248;
				    skE[0]&=127;
				    skE[0]|=64;
			    } else {
				    skE[ctx.ECDH.EGS-1]&=252;
				    skE[0]|=128;
			    }
		    }
	        
	        ctx.ECDH.KEY_PAIR_GENERATE(null, skE, pkE);
	        if (kem==32 || kem==33) 
		        this.reverse(pkR);

	        ctx.ECDH.ECPSVDP_DH(skE, pkR, DH);
	        if (kem==32 || kem==33) {
		        this.reverse(pkR);
		        this.reverse(pkE);
		        this.reverse(DH);
	        }
            var k=0;
            for (var i=0;i<pkE.length;i++)
                KEMCONTEXT[k++]=pkE[i];
            for (var i=0;i<pkR.length;i++)
                KEMCONTEXT[k++]=pkR[i];

    	    return this.ExtractAndExpand(DH,KEMCONTEXT);
        },

        decap: function(config_id,pkE,SKR) {
            var skR = [];
            var DH = [];
            var KEMCONTEXT=[];
            var pkR = [];
	        var kem = config_id&255;

	        for (var i=0;i<ctx.ECDH.EGS;i++)
		        skR[i]=SKR[i];
	        if (kem==32 || kem==33) {
		        this.reverse(skR);
		        this.reverse(pkE);
		        if (kem==32) {
			        skR[ctx.ECDH.EGS-1]&=248;
			        skR[0]&=127;
			        skR[0]|=64;
		        } else {
			        skR[ctx.ECDH.EGS-1]&=252;
			        skR[0]|=128;
		        }
	        }
	        ctx.ECDH.ECPSVDP_DH(skR, pkE, DH);
	        if (kem==32 || kem==33) {
		        this.reverse(pkE);
		        this.reverse(DH);
	        }
            ctx.ECDH.KEY_PAIR_GENERATE(null,skR,pkR);
	        if (kem==32 || kem==33) {
		        this.reverse(pkR);
	        }
            var k=0;
            for (var i=0;i<pkE.length;i++)
                KEMCONTEXT[k++]=pkE[i];
            for (var i=0;i<pkR.length;i++)
                KEMCONTEXT[k++]=pkR[i];

    	    return this.ExtractAndExpand(DH,KEMCONTEXT);
        },

        authEncap: function(config_id,SKE,pkE,pkR,SKS) {
            var pklen=pkE.length;
            var skE = [];
            var DH = [];
            var DH1 = [];
            var skS = [];
            var pkS = [];
            var KEMCONTEXT = [];
	
	        var kem = config_id&255;

		    for (var i=0;i<ctx.ECDH.EGS;i++) {
			    skE[i]=SKE[i];
			    skS[i]=SKS[i];
		    }	
		    if (kem==32 || kem==33) {
			    this.reverse(skE);
			    this.reverse(skS);
			    if (kem==32) {
				    skE[ctx.ECDH.EGS-1]&=248;
				    skE[0]&=127;
				    skE[0]|=64;
				    skS[ctx.ECDH.EGS-1]&=248;
				    skS[0]&=127;
				    skS[0]|=64;	
			    } else {
				    skE[ctx.ECDH.EGS-1]&=252;
				    skE[0]|=128;
				    skS[ctx.ECDH.EGS-1]&=252;
				    skS[0]|=128;
			    }
		    }
	     
	        ctx.ECDH.KEY_PAIR_GENERATE(null, skE, pkE);
	        if (kem==32 || kem==33) 
		        this.reverse(pkR);
	
	        ctx.ECDH.ECPSVDP_DH(skE, pkR, DH);
	        ctx.ECDH.ECPSVDP_DH(skS, pkR, DH1);

	        if (kem==32 || kem==33) {
		        this.reverse(DH);
		        this.reverse(DH1);
	        }
	        var ZZ = [];
	        for (var i=0;i<ctx.ECDH.EFS;i++) {
		        ZZ[i] = DH[i];
                ZZ[ctx.ECDH.EFS+i]= DH1[i];
	        }

            ctx.ECDH.KEY_PAIR_GENERATE(null, skS, pkS);

            if (kem==32 || kem==33)
            {
                this.reverse(pkR);
		        this.reverse(pkE);
                this.reverse(pkS);
            }
	        for (var i=0;i<pklen;i++) {
		        KEMCONTEXT[i] = pkE[i];
                KEMCONTEXT[pklen+i]= pkR[i];
                KEMCONTEXT[2*pklen+i]= pkS[i];
	        }

	        return this.ExtractAndExpand(ZZ,KEMCONTEXT);
        },

        authDecap: function(config_id,pkE,SKR,pkS) {
            var pklen=pkE.length;
            var skR = [];
            var DH = [];
            var DH1 = [];
            var pkR = [];
            var KEMCONTEXT = [];

		    var kem = config_id&255;

	        for (var i=0;i<ctx.ECDH.EGS;i++) 
		        skR[i]=SKR[i];
	
	        if (kem==32 || kem==33) {
		        this.reverse(skR);
		        this.reverse(pkE);
		        this.reverse(pkS);
		        if (kem==32) {
			        skR[ctx.ECDH.EGS-1]&=248;
			        skR[0]&=127;
			        skR[0]|=64;
		        } else {
		    	    skR[ctx.ECDH.EGS-1]&=252;
			        skR[0]|=128;
		        }  
	        }
	        ctx.ECDH.ECPSVDP_DH(skR, pkE, DH);
	        ctx.ECDH.ECPSVDP_DH(skR, pkS, DH1);

	        if (kem==32 || kem==33) {
		        this.reverse(DH);
		        this.reverse(DH1);
	        }
	        var ZZ = [];
	        for (var i=0;i<ctx.ECDH.EFS;i++) {
		        ZZ[i] = DH[i];
                ZZ[ctx.ECDH.EFS+i]= DH1[i];
	        }
            
            ctx.ECDH.KEY_PAIR_GENERATE(null, skR, pkR);

            if (kem==32 || kem==33)
            {
                this.reverse(pkR);
		        this.reverse(pkE);
                this.reverse(pkS);
            }
	        for (var i=0;i<pklen;i++) {
		        KEMCONTEXT[i] = pkE[i];
                KEMCONTEXT[pklen+i]= pkR[i];
                KEMCONTEXT[2*pklen+i]= pkS[i];
	        }

	        return this.ExtractAndExpand(ZZ,KEMCONTEXT);
        },

        keySchedule: function(config_id,key,nonce,exp_secret,mode,Z,info,psk,pskID) {
	        var context = [];
            var ZEROS = [];
	        var kem_id=config_id&255;
	        var kdf_id=(config_id>>8)&3;
	        var aead_id=(config_id>>10)&3;
            var k=0;
	
	        var ar=ctx.HMAC.inttobytes(kem_id, 2);
            for (var i=0;i<ar.length;i++)
		        context[k++] = ar[i];

	        ar=ctx.HMAC.inttobytes(kdf_id, 2);
            for (var i=0;i<ar.length;i++)
		        context[k++] = ar[i];

	        ar=ctx.HMAC.inttobytes(aead_id, 2);
            for (var i=0;i<ar.length;i++)
		        context[k++] = ar[i];

	        ar = ctx.HMAC.inttobytes(mode, 1);
            for (var i=0;i<ar.length;i++)
		        context[k++] = ar[i];

            for (var i=0;i<ctx.ECP.HASH_TYPE;i++)
                ZEROS[i]=0;

            var H=this.LabeledExtract(ZEROS,"pskID_hash",pskID);
            for (var i=0;i<ctx.ECP.HASH_TYPE;i++)
                context[k++]=H[i];
            H=this.LabeledExtract(ZEROS,"info",info);
            for (var i=0;i<ctx.ECP.HASH_TYPE;i++)
                context[k++]=H[i];

            if (psk==null)
                H=this.LabeledExtract(ZEROS,"psk_hash",ZEROS);
            else
                H=this.LabeledExtract(ZEROS,"psk_hash",psk);
            var secret=this.LabeledExtract(H,"zz",Z);

            var ex=this.LabeledExpand(secret,"key",context,ctx.ECP.AESKEY);
            for (var i=0;i<ex.length;i++) key[i]=ex[i];

            ex=this.LabeledExpand(secret,"nonce",context,12);
            for (var i=0;i<ex.length;i++) nonce[i]=ex[i];

            if (exp_secret!=null)
            {
                ex=this.LabeledExpand(secret,"exp",context,ctx.ECP.HASH_TYPE);
                for (var i=0;i<ex.length;i++) exp_secret[i]=ex[i];
            }
        }

    };
    return HPKE;
};
